### PR TITLE
Add support for multi-column subqueries in IN predicates and quantified comparisons

### LINKED ISSUE #7797: Incorrect handling of row expressions in IN predicate, quantified comparisons and scalar subqueries
Currently, the following queries returns `true`:

```sql
SELECT ROW(1, 2) IN (SELECT ROW(1, 2))

SELECT ROW(1, 2) = ALL (SELECT ROW(1, 2))
```

This is incorrect because according to the SQL specification:

```
<quantified comparison predicate> ::=
  <row value predicand> <quantified comparison predicate part 2>
<quantified comparison predicate part 2> ::=
  <comp op> <quantifier> <table subquery>

1)  Let R be the result of the <row value predicand> and let T be the result of the <table subquery>.
2) The result of “R <comp op> <quantifier> T” is derived by the application of the implied 
    <comparison predicate> “R <comp op> RT” to every row RT in T.

7) The declared types of the columns of a <table subquery> are the declared types of the respective columns of QE.
```

So, the type of the subquery is a table containing a *single* column of type `row(integer, integer)`. Since the left side is a `row` with two fields, it's incompatible with the type of the table. The queries should fail with a type mismatch error.

In particular:

```sql
SELECT ROW(1, 2) IN (SELECT 1, 2)
```

should be equivalent to:

```sql
SELECT ROW(1, 2) IN (ROW(1, 2))
```

so,

```sql
SELECT ROW(1, 2) IN (SELECT ROW(1, 2))
```

should be equivalent to this query, which makes the type mismatch more evident:

```sql
SELECT ROW(1, 2) IN (ROW(ROW(1, 2)))
```

### PR DESCRIPTION
**Pull request by martint**
Also, fixes https://github.com/trinodb/trino/issues/7797

### CODE REVIEW

**Review by kasiafi — COMMENTED** [review]
Some questions / comments.
I haven't gone through `SubqueryPlanner` yet.

**Comment by kasiafi** [comment]
I've read the `SubqueryPlanner` part. It looks good. It is very neat and clear as for Planner :)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalysis.java [hunk 1]
```diff
@@ -98,6 +102,11 @@ public boolean isColumnReference(Expression node)
         return subqueries;
     }
 
+    public Map<NodeRef<Expression>, PredicateWithSubqueryAnalysis> getPredicateWithSubqueryAnalysis()
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalysis.java:105]: No, it isn't. I'll remove it

File: core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java [hunk 2]
```diff
@@ -1329,114 +1329,257 @@ public void testQuantifiedComparison()
...
+
+        assertThat(assertions.query(
+                "SELECT ROW(BIGINT '1', BIGINT '2') = ALL (SELECT 1, 2)"))
+                .describedAs("Multi-column row type with coercion on subquery side")
+                .matches("VALUES true");
+
+        assertThat(assertions.query(
+                "SELECT ROW(BIGINT '1', 2) = ALL (SELECT 1, BIGINT '2')"))
+                .describedAs("Multi-column row type with coercion on both sides")
+                .matches("VALUES true");
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1423]: The latter. Null is a value of every type

[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1423]: Actually, let me look at the spec. Currently, the code fails for that. So it's either a bug in the code or expected behavior.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 3]
```diff
@@ -1362,6 +1373,38 @@ public void addQuantifiedComparisons(List<QuantifiedComparisonExpression> expres
         }
     }
 
+    /**
+     * Analysis for predicates such as <code>x IN (subquery)</code> or <code>x = SOME (subquery)</code>
+     */
+    public static class PredicateWithSubqueryAnalysis
+    {
+        private final Type actualValueType; // the type of the value-side of the predicate after the value to single-element-row conversion has been done
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:1381]: Thanks for the suggestion. Yes, I agree. I'm changing the name and type to make the meaning clearer.

```java
        private final RowType valueType;
        private final Optional<RowType> valueCoercion;
        private final Optional<RowType> subqueryCoercion;
```

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -1909,17 +1918,23 @@ private Type analyzeSubquery(SubqueryExpression node, StackableAstVisitorContext
...
+                        fields.add(RowType.field(field.getName().get(), field.getType()));
+                    }
+                    else {
+                        fields.add(RowType.field(field.getType()));
+                    }
+                }
             }
 
-            sourceFields.add(queryScope.getRelationType().getFieldByIndex(0));
+            Type type = RowType.from(fields.build());
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1934]: That can't happen. There's no way to select no fields from a table (i.e., you can't do `SELECT FROM t`). If there are no visible fields, it's a bug, and it's ok for that constructor to fail.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -1856,46 +1862,49 @@ protected Type visitInListExpression(InListExpression node, StackableAstVisitorC
         @Override
         protected Type visitSubqueryExpression(SubqueryExpression node, StackableAstVisitorContext<Context> context)
         {
-            Type type = analyzeScalarSubquery(node, context);
+            Type type = analyzeSubquery(node, context);
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1865]: No, this also handles the case for `row subqueries`. E.g., `SELECT x  = (SELECT 1, 2 FROM t)`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 6]
```diff
@@ -1957,13 +1972,8 @@ protected Type visitQuantifiedComparisonExpression(QuantifiedComparisonExpressio
...
 
-            Expression value = node.getValue();
-            process(value, context);
-
-            Expression subquery = node.getSubquery();
-            analyzeSubquery((SubqueryExpression) subquery, context);
-
-            Type comparisonType = coerceToSingleType(context, node, "Value expression and result of subquery must be of the same type for quantified comparison: %s vs %s", value, subquery);
+            Type declaredValueType = process(node.getValue(), context);
+            Type comparisonType = analyzePredicateWithSubquery(node, declaredValueType, (SubqueryExpression) node.getSubquery(), context);
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1976]: That's more complicated and it's not what we do today. Let's make that a future improvement.

File: core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1423]: The null is a `<contextually typed value specification>`:

> The declared type DT of a <null specification> NS is determined by the context in which NS appears. NS
is effectively replaced by CAST ( NS AS DT ).

So, the type is supposed to be inferred based on the type of the other side.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1934]: Actually, there's a (preexisting) issue here. Technically, the type of `VALUES null` is `multiset(unknown)`. There's no way to know what the actual type of those rows is without looking at where the result of that VALUES construct is used (it's a "contextually defined xxx", as I mentioned elsewhere). Currently, due to how Scope and RelationType works, there's no way to represent anything other that a concrete row type, so the type of `VALUES null` would appear as a relation with no columns.

To fix this, I need to fix the analysis of `VALUES`. Let's see what other things need to be adjusted. I'll add a separate commit (or possibly PR) for that.

[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1934]: It turns out I misread that part of the spec. `VALUES null` is of type `VALUES ROW(null)`:

> If CTRVC is a `<common value expression>`, `<boolean value expression>`, or `<contextually typed value
specification>` X, then CTRVC is equivalent to: ROW ( X )

File: core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1423]: The answer is:

`SELECT null = ALL (SELECT 1, 2)` is equivalent to `SELECT CAST(null AS row(integer, integer)) = ALL (SELECT 1, 2)`

`SELECT ROW(1, 2) IN (SELECT null)` is equivalent to `SELECT ROW(1, 2) IN (VALUES ROW(null))`

**Comment by martint** [comment]
@kasiafi , addressed the latest comments in a fixup commit

File: core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java [hunk 7]
```diff
@@ -411,40 +405,116 @@ private PlanBuilder addNegation(PlanBuilder subPlan, Cluster<? extends Expressio
...
+            Symbol wrapped = symbolAllocator.newSymbol("row", actualType);
+
+            Assignments assignments = Assignments.builder()
+                    .putIdentities(subPlan.getRoot().getOutputSymbols())
+                    .put(wrapped, new Row(ImmutableList.of(column.toSymbolReference())))
+                    .build();
+
+            subPlan = new PlanBuilder(
+                    subPlan.getTranslations(),
+                    new ProjectNode(idAllocator.getNextId(), subPlan.getRoot(), assignments));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:450]: `subPlan = subPlan.withNewRoot(new ProjectNode(idAllocator.getNextId(), subPlan.getRoot(), assignments));`

**Review by kasiafi — APPROVED** [review]
Thanks. Squash the commits please.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 8]
```diff
@@ -1834,8 +1842,7 @@ protected Type visitInPredicate(InPredicate node, StackableAstVisitorContext<Con
             }
             else if (valueList instanceof SubqueryExpression) {
                 subqueryInPredicates.add(NodeRef.of(node));
-                analyzeSubquery((SubqueryExpression) valueList, context);
-                coerceToSingleType(context, node, "value and result of subquery must be of the same type for IN expression: %s vs %s", value, valueList);
+                analyzePredicateWithSubquery(node, declaredValueType, (SubqueryExpression) valueList, context);
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1845]: I noticed that, too. The original code wasn't doing any checks. I'll fix that in a separate PR.

**Pull request merged**